File: assignment1_rt_py/assignment1_rt_py/distance_node.py
#!/usr/bin/env python3
import math
import time

import rclpy
from rclpy.node import Node

from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from std_msgs.msg import Bool

from turtlesim.srv import SetPen, TeleportAbsolute

# ANSI
Y = "\033[33m"
R = "\033[0m"

# ================================
# CONFIG
# ================================
PRECOLLISION_DISTANCE = 1.5
COLLISION_DISTANCE = 0.5

WALL_THRESHOLD_MIN = 0.5
WALL_THRESHOLD_MAX = 10.5

SLEEP_AFTER_COLLISION = 0.4  # seconds


class DistanceNode(Node):

    def __init__(self):
        super().__init__("distance_node")

        self.get_logger().info(f"{Y}distance_node 2.3 patched (micro-move) loaded{R}")

        # Internal
        self.pose1 = None
        self.pose2 = None

        self.last_stop1 = None
        self.last_stop2 = None

        self.init1 = False
        self.init2 = False

        self.pre1_active = False
        self.pre2_active = False

        # Subscribers
        self.create_subscription(Pose, "/turtle1/pose", self.pose1_callback, 10)
        self.create_subscription(Pose, "/turtle2/pose", self.pose2_callback, 10)

        # Publishers
        self.pub1 = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.pub2 = self.create_publisher(Twist, "/turtle2/cmd_vel", 10)
        self.freeze_pub = self.create_publisher(Bool, "/freeze_turtles", 10)

        # Services
        self.pen1 = self.create_client(SetPen, "/turtle1/set_pen")
        self.pen2 = self.create_client(SetPen, "/turtle2/set_pen")
        self.tp1 = self.create_client(TeleportAbsolute, "/turtle1/teleport_absolute")
        self.tp2 = self.create_client(TeleportAbsolute, "/turtle2/teleport_absolute")

        # Timer
        self.create_timer(0.05, self.update)

    # -----------------------------
    # Utility
    # -----------------------------
    def freeze(self, state: bool):
        msg = Bool()
        msg.data = state
        self.freeze_pub.publish(msg)

    def stop_turtle(self, pub):
        msg = Twist()
        msg.linear.x = 0.0
        msg.angular.z = 0.0
        pub.publish(msg)

    def set_pen(self, client, r, g, b, w, off):
        if not client.wait_for_service(timeout_sec=0.1):
            return
        req = SetPen.Request()
        req.r, req.g, req.b = r, g, b
        req.width = w
        req.off = 1 if off else 0
        client.call_async(req)

    def teleport(self, client, pose):
        if pose is None or not client.wait_for_service(timeout_sec=0.1):
            return
        req = TeleportAbsolute.Request()
        req.x = pose.x
        req.y = pose.y
        req.theta = pose.theta
        client.call_async(req)

    @staticmethod
    def near_wall(p: Pose):
        return (
            p.x < WALL_THRESHOLD_MIN
            or p.x > WALL_THRESHOLD_MAX
            or p.y < WALL_THRESHOLD_MIN
            or p.y > WALL_THRESHOLD_MAX
        )

    # -----------------------------
    # Pose callbacks
    # -----------------------------
    def pose1_callback(self, msg):
        self.pose1 = msg
        if not self.init1:
            self.init1 = True
            self.last_stop1 = msg
            self.set_pen(self.pen1, 0, 0, 255, 3, False)
        if msg.linear_velocity == 0.0 and msg.angular_velocity == 0.0:
            self.last_stop1 = msg

    def pose2_callback(self, msg):
        self.pose2 = msg
        if not self.init2:
            self.init2 = True
            self.last_stop2 = msg
            self.set_pen(self.pen2, 0, 0, 255, 3, False)
        if msg.linear_velocity == 0.0 and msg.angular_velocity == 0.0:
            self.last_stop2 = msg

    # -----------------------------
    # Main loop
    # -----------------------------
    def update(self):
        if self.pose1 is None or self.pose2 is None:
            return

        dx = self.pose1.x - self.pose2.x
        dy = self.pose1.y - self.pose2.y
        dist = math.sqrt(dx * dx + dy * dy)

        # Pre-collision pen logic
        self.handle_pre_collision(dist)

        # Collision
        if dist < COLLISION_DISTANCE:
            self.handle_collision(both=True)
            return

        # Wall collisions
        if self.near_wall(self.pose1):
            self.handle_collision(t1=True)
            return

        if self.near_wall(self.pose2):
            self.handle_collision(t2=True)
            return

    # -----------------------------
    # Pre-collision (pen red)
    # -----------------------------
    def handle_pre_collision(self, dist):
        in_pre = dist < PRECOLLISION_DISTANCE

        if in_pre and not self.pre1_active:
            self.set_pen(self.pen1, 255, 0, 0, 3, False)
            self.pre1_active = True
        if not in_pre and self.pre1_active:
            self.set_pen(self.pen1, 0, 0, 255, 3, False)
            self.pre1_active = False

        if in_pre and not self.pre2_active:
            self.set_pen(self.pen2, 255, 0, 0, 3, False)
            self.pre2_active = True
        if not in_pre and self.pre2_active:
            self.set_pen(self.pen2, 0, 0, 255, 3, False)
            self.pre2_active = False

    # -----------------------------
    # Collision handling (patched)
    # -----------------------------
    def handle_collision(self, both=False, t1=False, t2=False):
        self.freeze(True)

        # Select turtles
        use1 = both or t1
        use2 = both or t2

        # Micro movement message
        micro = Twist()
        micro.linear.x = 1.5
        micro.angular.z = 0.0

        # 1. Activate red pen + micro-move
        if use1:
            self.set_pen(self.pen1, 255, 0, 0, 4, False)
            self.pub1.publish(micro)
        if use2:
            self.set_pen(self.pen2, 255, 0, 0, 4, False)
            self.pub2.publish(micro)

        time.sleep(0.12)  # controls minimal dash length

        # 2. Stop & disable pen
        if use1:
            self.stop_turtle(self.pub1)
            self.set_pen(self.pen1, 0, 0, 0, 4, True)

        if use2:
            self.stop_turtle(self.pub2)
            self.set_pen(self.pen2, 0, 0, 0, 4, True)

        # 3. Teleport
        if use1:
            self.teleport(self.tp1, self.last_stop1)
        if use2:
            self.teleport(self.tp2, self.last_stop2)

        time.sleep(0.05)

        # 4. Restore blue
        if use1:
            self.set_pen(self.pen1, 0, 0, 255, 3, False)
        if use2:
            self.set_pen(self.pen2, 0, 0, 255, 3, False)

        self.freeze(False)


# -----------------------------
# Main
# -----------------------------
def main(args=None):
    rclpy.init(args=args)
    node = DistanceNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: assignment1_rt_cpp/src/distance_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "std_msgs/msg/bool.hpp"
#include "turtlesim/srv/set_pen.hpp"
#include "turtlesim/srv/teleport_absolute.hpp"

#include <cmath>
#include <chrono>
#include <memory>
#include <string>
#include <functional>

// -------------------------
// CONFIG
// -------------------------
static const float PRECOLLISION_DISTANCE = 1.5f;
static const float COLLISION_DISTANCE   = 0.5f;

static const float WALL_MIN = 0.5f;
static const float WALL_MAX = 10.5f;

static const int COLLISION_SLEEP_MS = 400;

// -------------------------
using std::string;
using std::bind;
using std::placeholders::_1;
using std::sqrt;
using std::chrono::milliseconds;

using rclcpp::Node;
using rclcpp::sleep_for;
using rclcpp::init;
using rclcpp::spin;
using rclcpp::shutdown;

template<typename T> using Sub = rclcpp::Subscription<T>;
template<typename T> using Pub = rclcpp::Publisher<T>;
using Timer = rclcpp::TimerBase;

using Pose = turtlesim::msg::Pose;
using Twist = geometry_msgs::msg::Twist;
using BoolMsg = std_msgs::msg::Bool;

using SetPenSrv = turtlesim::srv::SetPen;
using TeleportSrv = turtlesim::srv::TeleportAbsolute;

using ClientSetPen = rclcpp::Client<SetPenSrv>;
using ClientTeleport = rclcpp::Client<TeleportSrv>;

using SetPenReq = SetPenSrv::Request;
using TeleportReq = TeleportSrv::Request;

// ANSI Colors
static const string Y = "\033[33m";
static const string R = "\033[0m";


// =====================================================
// NODE
// =====================================================
class DistanceNode : public Node
{
public:
    DistanceNode() : Node("distance_node")
    {
        RCLCPP_INFO(
            this->get_logger(),
            "%sdistance_node v5.1 loaded%s",
            Y.c_str(), R.c_str()
        );

        sub1_ = create_subscription<Pose>(
            "/turtle1/pose", 10,
            bind(&DistanceNode::pose1Callback, this, _1));

        sub2_ = create_subscription<Pose>(
            "/turtle2/pose", 10,
            bind(&DistanceNode::pose2Callback, this, _1));

        pub1_ = create_publisher<Twist>("/turtle1/cmd_vel", 10);
        pub2_ = create_publisher<Twist>("/turtle2/cmd_vel", 10);
        freeze_pub_ = create_publisher<BoolMsg>("/freeze_turtles", 10);

        pen1_client_ = create_client<SetPenSrv>("/turtle1/set_pen");
        pen2_client_ = create_client<SetPenSrv>("/turtle2/set_pen");

        tp1_client_ = create_client<TeleportSrv>("/turtle1/teleport_absolute");
        tp2_client_ = create_client<TeleportSrv>("/turtle2/teleport_absolute");

        timer_ = create_wall_timer(
            milliseconds(50),
            bind(&DistanceNode::update, this));
    }

private:

    void setFreeze(bool s)
    {
        BoolMsg msg;
        msg.data = s;
        freeze_pub_->publish(msg);
    }

    void setPen(const ClientSetPen::SharedPtr &client,
                int r, int g, int b, int w, bool off)
    {
        if (!client->wait_for_service(milliseconds(100)))
            return;

        auto req = std::make_shared<SetPenReq>();
        req->r = r; req->g = g; req->b = b;
        req->width = w;
        req->off = off ? 1 : 0;

        client->async_send_request(req);
    }

    void teleport(const ClientTeleport::SharedPtr &client, const Pose &p)
    {
        if (!client->wait_for_service(milliseconds(100)))
            return;

        auto req = std::make_shared<TeleportReq>();
        req->x = p.x;
        req->y = p.y;
        req->theta = p.theta;

        client->async_send_request(req);
    }

    void stopTurtle(const Pub<Twist>::SharedPtr &pub)
    {
        Twist msg;
        msg.linear.x  = 0;
        msg.angular.z = 0;
        pub->publish(msg);
    }

    bool isNearWall(const Pose &p)
    {
        return (p.x < WALL_MIN || p.x > WALL_MAX ||
                p.y < WALL_MIN || p.y > WALL_MAX);
    }

    // =====================================================
    // CALLBACKS
    // =====================================================
    void pose1Callback(const Pose &p)
    {
        pose1_ = p;
        pose1_ready_ = true;

        if (!init1_) {
            last_stop1_ = p;
            init1_ = true;
            setPen(pen1_client_, 0, 0, 255, 3, false);
        }

        if (p.linear_velocity == 0 && p.angular_velocity == 0)
            last_stop1_ = p;
    }

    void pose2Callback(const Pose &p)
    {
        pose2_ = p;
        pose2_ready_ = true;

        if (!init2_) {
            last_stop2_ = p;
            init2_ = true;
            setPen(pen2_client_, 0, 0, 255, 3, false);
        }

        if (p.linear_velocity == 0 && p.angular_velocity == 0)
            last_stop2_ = p;
    }

    // =====================================================
    // MAIN LOOP
    // =====================================================
    void update()
    {
        if (!pose1_ready_ || !pose2_ready_)
            return;

        float dx   = pose1_.x - pose2_.x;
        float dy   = pose1_.y - pose2_.y;
        float dist = sqrt(dx * dx + dy * dy);

        handlePreCollision(dist);

        if (dist < COLLISION_DISTANCE) {
            handleCollision(true, true);
            return;
        }

        if (isNearWall(pose1_)) {
            handleCollision(true, false);
            return;
        }

        if (isNearWall(pose2_)) {
            handleCollision(false, true);
            return;
        }
    }

    // =====================================================
    // PRE COLLISION
    // =====================================================
    void handlePreCollision(float dist)
    {
        bool in_pre = (dist < PRECOLLISION_DISTANCE);

        if (in_pre && !pre1_active_) {
            setPen(pen1_client_, 255, 0, 0, 3, false);
            pre1_active_ = true;
        }
        if (!in_pre && pre1_active_) {
            setPen(pen1_client_, 0, 0, 255, 3, false);
            pre1_active_ = false;
        }

        if (in_pre && !pre2_active_) {
            setPen(pen2_client_, 255, 0, 0, 3, false);
            pre2_active_ = true;
        }
        if (!in_pre && pre2_active_) {
            setPen(pen2_client_, 0, 0, 255, 3, false);
            pre2_active_ = false;
        }
    }

    // =====================================================
    // COLLISION (PATCHED WITH RED MICRO-MOVE)
    // =====================================================
    void handleCollision(bool t1, bool t2)
    {
        setFreeze(true);

        // Micro-move twist
        Twist micro;
        micro.linear.x = 1.5;
        micro.angular.z = 0.0;

        // 1) Red pen + micro-move = guaranteed red dash
        if (t1) setPen(pen1_client_, 255, 0, 0, 4, false);
        if (t2) setPen(pen2_client_, 255, 0, 0, 4, false);

        if (t1) pub1_->publish(micro);
        if (t2) pub2_->publish(micro);

        sleep_for(milliseconds(120)); // red dash duration

        // 2) Stop movement and disable pen (no teleport streaks)
        if (t1) stopTurtle(pub1_);
        if (t2) stopTurtle(pub2_);

        if (t1) setPen(pen1_client_, 0, 0, 0, 4, true);
        if (t2) setPen(pen2_client_, 0, 0, 0, 4, true);

        // 3) Teleport
        if (t1) teleport(tp1_client_, last_stop1_);
        if (t2) teleport(tp2_client_, last_stop2_);

        sleep_for(milliseconds(50));

        // 4) Restore blue
        if (t1) setPen(pen1_client_, 0, 0, 255, 3, false);
        if (t2) setPen(pen2_client_, 0, 0, 255, 3, false);

        setFreeze(false);
    }


    // =====================================================
    // VARIABLES
    // =====================================================
    Pose pose1_, pose2_;
    Pose last_stop1_, last_stop2_;

    bool pose1_ready_ = false;
    bool pose2_ready_ = false;
    bool init1_ = false;
    bool init2_ = false;

    bool pre1_active_ = false;
    bool pre2_active_ = false;

    Sub<Pose>::SharedPtr sub1_, sub2_;
    Pub<Twist>::SharedPtr pub1_, pub2_;
    Pub<BoolMsg>::SharedPtr freeze_pub_;

    ClientSetPen::SharedPtr pen1_client_, pen2_client_;
    ClientTeleport::SharedPtr tp1_client_, tp2_client_;

    Timer::SharedPtr timer_;
};

// =====================================================
int main(int argc, char **argv)
{
    init(argc, argv);
    spin(std::make_shared<DistanceNode>());
    shutdown();
    return 0;
}




